File: server/watcher.py
"""File watcher for automatic reindexing."""

import hashlib
import os
import time
import logging
import threading
from pathlib import Path
from watchdog.observers import Observer
from watchdog.events import FileSystemEventHandler

from .db import GraphDB
from .parser import SpaceParser
from .config_parser import parse_config_page, write_config_json


class SpaceWatcher(FileSystemEventHandler):
    """Handle file system events in the Silverbullet space."""

    def __init__(
        self,
        space_path: str,
        graph_db: GraphDB = None,
        parser: SpaceParser = None,
        db_path: str = None,
    ):
        self.space_path = space_path
        self.graph_db = graph_db if graph_db else GraphDB("/db")
        self.parser = parser if parser else SpaceParser()
        self.db_path = db_path or os.getenv("DB_PATH", "/data/ladybug")
        self.debounce_time = {}
        self.processing_lock = threading.Lock()
        self.currently_processing = set()
        self.file_hashes = {}  # Track content hashes to avoid unnecessary reindexing

    def _compute_file_hash(self, path: str) -> str | None:
        """Compute MD5 hash of file contents.

        Args:
            path: Path to the file

        Returns:
            MD5 hash string, or None if file cannot be read
        """
        try:
            with open(path, "rb") as f:
                return hashlib.md5(f.read()).hexdigest()
        except Exception as e:
            logging.debug(f"Could not hash {path}: {e}")
            return None

    def _has_content_changed(self, path: str) -> bool:
        """Check if file content has actually changed since last indexing.

        Args:
            path: Path to the file

        Returns:
            True if content has changed or file is new, False otherwise
        """
        current_hash = self._compute_file_hash(path)
        if current_hash is None:
            return True  # Assume changed if we can't read it

        previous_hash = self.file_hashes.get(path)
        if previous_hash is None:
            # New file, hasn't been indexed before
            return True

        if current_hash == previous_hash:
            logging.debug(
                f"Skipping {path} - content unchanged (hash: {current_hash[:8]}...)"
            )
            return False

        return True

    def _update_file_hash(self, path: str):
        """Update the stored hash for a file after successful indexing."""
        current_hash = self._compute_file_hash(path)
        if current_hash:
            self.file_hashes[path] = current_hash

    def _should_process(self, path: str) -> bool:
        """Check if file should be processed (debouncing, deduplication, and content check)."""
        now = time.time()
        last_time = self.debounce_time.get(path, 0)

        # Debounce: only process if 5 seconds have passed since last processing
        if now - last_time < 5.0:
            return False

        # Check if already being processed
        with self.processing_lock:
            if path in self.currently_processing:
                logging.debug(f"Skipping {path} - already being processed")
                return False

        # Check if content has actually changed
        if not self._has_content_changed(path):
            # Update debounce time even for unchanged files to prevent repeated checks
            self.debounce_time[path] = now
            return False

        return True

    def _mark_processing(self, path: str) -> bool:
        """Mark a file as being processed. Returns False if already processing."""
        with self.processing_lock:
            if path in self.currently_processing:
                return False
            self.currently_processing.add(path)
            self.debounce_time[path] = time.time()
            return True

    def _unmark_processing(self, path: str):
        """Unmark a file as being processed."""
        with self.processing_lock:
            self.currently_processing.discard(path)

    def on_modified(self, event):
        """Handle file modification."""
        if event.is_directory or not event.src_path.endswith(".md"):
            return

        if not self._should_process(event.src_path):
            return

        logging.info(f"File modified: {event.src_path}")
        self._reindex_file(event.src_path)

    def on_created(self, event):
        """Handle file creation."""
        if event.is_directory or not event.src_path.endswith(".md"):
            return

        if not self._should_process(event.src_path):
            return

        logging.info(f"File created: {event.src_path}")
        self._reindex_file(event.src_path)

    def on_deleted(self, event):
        """Handle file deletion."""
        if event.is_directory or not event.src_path.endswith(".md"):
            return

        logging.info(f"File deleted: {event.src_path}")

        # Remove deleted file's chunks from graph
        try:
            self.graph_db.delete_chunks_by_file(event.src_path)
            logging.info(f"Deleted chunks for {event.src_path}")
        except Exception as e:
            logging.error(f"Error deleting chunks: {e}")

    def _handle_config_change(self, file_path: str) -> None:
        """Parse CONFIG.md and write space_config.json.

        This extracts config.set() values from space-lua blocks and writes
        them to a JSON file that the MCP server can read.

        Args:
            file_path: Path to the CONFIG.md file
        """
        try:
            content = Path(file_path).read_text(encoding="utf-8")
            config = parse_config_page(content)
            db_path = Path(self.db_path)
            write_config_json(config, db_path)
            logging.info(f"Updated space_config.json from {file_path}")
        except Exception as e:
            logging.error(f"Failed to parse CONFIG.md: {e}")

    def _reindex_file(self, file_path: str):
        """Reindex a single file instead of the entire space."""
        if not self._mark_processing(file_path):
            logging.debug(f"Skipping {file_path} - already being processed")
            return

        try:
            # Special handling for CONFIG.md - parse config values
            if file_path.endswith("CONFIG.md"):
                self._handle_config_change(file_path)

            # Delete existing chunks for this file
            self.graph_db.delete_chunks_by_file(file_path)

            # Parse only the changed file
            chunks = self.parser.parse_file(file_path)
            if chunks:
                self.graph_db.index_chunks(chunks)
                logging.info(f"Reindexed {len(chunks)} chunks from {file_path}")
            else:
                logging.info(f"No chunks found in {file_path}")

            # Update hash after successful indexing
            self._update_file_hash(file_path)
        except Exception as e:
            logging.error(f"Reindexing error for {file_path}: {e}")
        finally:
            self._unmark_processing(file_path)

    def reindex_all(self):
        """Reindex the entire space. Use for initial indexing."""
        try:
            chunks = self.parser.parse_space(self.space_path)
            self.graph_db.index_chunks(chunks)
            logging.info(f"Reindexed {len(chunks)} chunks from entire space")

            # Build initial hash cache for all markdown files
            space_path = Path(self.space_path)
            for md_file in space_path.glob("**/*.md"):
                self._update_file_hash(str(md_file))
            logging.info(f"Cached hashes for {len(self.file_hashes)} files")
        except Exception as e:
            logging.error(f"Full reindexing error: {e}")


def watch_space(space_path: str = "/space"):
    """Start watching the Silverbullet space for changes.

    Args:
        space_path: Path to the space directory
    """
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s - %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )

    event_handler = SpaceWatcher(space_path)
    observer = Observer()
    observer.schedule(event_handler, space_path, recursive=True)

    logging.info(f"Starting file watcher on {space_path}")
    observer.start()

    try:
        while True:
            time.sleep(1)
    except KeyboardInterrupt:
        observer.stop()
        logging.info("File watcher stopped")

    observer.join()


# Alias for backward compatibility
SpaceEventHandler = SpaceWatcher


if __name__ == "__main__":
    watch_space()

File: internal/watcher/watcher.go
// Package watcher provides file system watching for SilverBullet spaces.
package watcher

import (
	"context"
	"crypto/md5"
	"encoding/hex"
	"io"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/fsnotify/fsnotify"

	"github.com/boblangley/silverbullet-rag/internal/config"
	"github.com/boblangley/silverbullet-rag/internal/db"
	"github.com/boblangley/silverbullet-rag/internal/embeddings"
	"github.com/boblangley/silverbullet-rag/internal/parser"
	"github.com/boblangley/silverbullet-rag/internal/types"
)

// Watcher watches a SilverBullet space for file changes.
type Watcher struct {
	spacePath  string
	db         *db.GraphDB
	parser     *parser.SpaceParser
	embedding  *embeddings.Service
	denoRunner *config.DenoRunner
	dbPath     string
	logger     *slog.Logger

	watcher  *fsnotify.Watcher
	debounce time.Duration
	pending  map[string]time.Time
	mu       sync.Mutex

	// Hash tracking to avoid reprocessing unchanged files
	fileHashes map[string]string
	hashMu     sync.RWMutex

	// Concurrent processing tracking
	currentlyProcessing map[string]bool
	processingMu        sync.Mutex
}

// Config holds watcher configuration.
type Config struct {
	SpacePath  string
	DB         *db.GraphDB
	Embedding  *embeddings.Service
	DenoRunner *config.DenoRunner
	DBPath     string
	Debounce   time.Duration
	Logger     *slog.Logger
}

// New creates a new file watcher.
func New(cfg Config) (*Watcher, error) {
	fsWatcher, err := fsnotify.NewWatcher()
	if err != nil {
		return nil, err
	}

	debounce := cfg.Debounce
	if debounce == 0 {
		debounce = 500 * time.Millisecond
	}

	logger := cfg.Logger
	if logger == nil {
		logger = slog.Default()
	}

	return &Watcher{
		spacePath:           cfg.SpacePath,
		db:                  cfg.DB,
		parser:              parser.NewSpaceParser(cfg.SpacePath),
		embedding:           cfg.Embedding,
		denoRunner:          cfg.DenoRunner,
		dbPath:              cfg.DBPath,
		logger:              logger,
		watcher:             fsWatcher,
		debounce:            debounce,
		pending:             make(map[string]time.Time),
		fileHashes:          make(map[string]string),
		currentlyProcessing: make(map[string]bool),
	}, nil
}

// Start begins watching the space directory.
func (w *Watcher) Start(ctx context.Context) error {
	// Add all directories recursively
	err := filepath.WalkDir(w.spacePath, func(path string, d os.DirEntry, err error) error {
		if err != nil {
			return err
		}

		if d.IsDir() {
			// Skip hidden directories
			if strings.HasPrefix(d.Name(), ".") {
				return filepath.SkipDir
			}
			return w.watcher.Add(path)
		}
		return nil
	})
	if err != nil {
		return err
	}

	w.logger.Info("started watching space", "path", w.spacePath)

	// Start event processing
	go w.processEvents(ctx)

	// Start debounce processor
	go w.processDebounced(ctx)

	return nil
}

// Stop stops the watcher.
func (w *Watcher) Stop() error {
	return w.watcher.Close()
}

func (w *Watcher) processEvents(ctx context.Context) {
	for {
		select {
		case <-ctx.Done():
			return

		case event, ok := <-w.watcher.Events:
			if !ok {
				return
			}

			// Only handle markdown files
			if !strings.HasSuffix(event.Name, ".md") {
				// Handle new directories
				if event.Op&fsnotify.Create != 0 {
					if info, err := os.Stat(event.Name); err == nil && info.IsDir() {
						_ = w.watcher.Add(event.Name)
					}
				}
				continue
			}

			// Skip proposal files
			if strings.HasSuffix(event.Name, ".proposal") ||
				strings.Contains(event.Name, "_Proposals") ||
				strings.HasSuffix(event.Name, ".rejected.md") {
				continue
			}

			// Queue for debounced processing
			w.mu.Lock()
			w.pending[event.Name] = time.Now()
			w.mu.Unlock()

		case err, ok := <-w.watcher.Errors:
			if !ok {
				return
			}
			w.logger.Error("watcher error", "error", err)
		}
	}
}

func (w *Watcher) processDebounced(ctx context.Context) {
	ticker := time.NewTicker(100 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ctx.Done():
			return

		case <-ticker.C:
			w.mu.Lock()
			now := time.Now()
			var ready []string

			for path, queueTime := range w.pending {
				if now.Sub(queueTime) >= w.debounce {
					ready = append(ready, path)
				}
			}

			for _, path := range ready {
				delete(w.pending, path)
			}
			w.mu.Unlock()

			// Process ready files
			for _, path := range ready {
				w.handleFileChange(ctx, path)
			}
		}
	}
}

// computeFileHash computes MD5 hash of file contents.
func (w *Watcher) computeFileHash(filePath string) (string, error) {
	f, err := os.Open(filePath)
	if err != nil {
		return "", err
	}
	defer f.Close()

	h := md5.New()
	if _, err := io.Copy(h, f); err != nil {
		return "", err
	}

	return hex.EncodeToString(h.Sum(nil)), nil
}

// hasContentChanged checks if file content has changed since last index.
func (w *Watcher) hasContentChanged(filePath string) bool {
	currentHash, err := w.computeFileHash(filePath)
	if err != nil {
		// If we can't compute hash, assume changed
		return true
	}

	w.hashMu.RLock()
	storedHash, exists := w.fileHashes[filePath]
	w.hashMu.RUnlock()

	if !exists {
		return true
	}

	return currentHash != storedHash
}

// updateFileHash stores the current hash for a file.
func (w *Watcher) updateFileHash(filePath string) {
	hash, err := w.computeFileHash(filePath)
	if err != nil {
		return
	}

	w.hashMu.Lock()
	w.fileHashes[filePath] = hash
	w.hashMu.Unlock()
}

// clearFileHash removes the stored hash for a file.
func (w *Watcher) clearFileHash(filePath string) {
	w.hashMu.Lock()
	delete(w.fileHashes, filePath)
	w.hashMu.Unlock()
}

// markProcessing marks a file as currently being processed.
// Returns false if already being processed.
func (w *Watcher) markProcessing(filePath string) bool {
	w.processingMu.Lock()
	defer w.processingMu.Unlock()

	if w.currentlyProcessing[filePath] {
		return false
	}

	w.currentlyProcessing[filePath] = true
	return true
}

// unmarkProcessing removes a file from the processing set.
func (w *Watcher) unmarkProcessing(filePath string) {
	w.processingMu.Lock()
	delete(w.currentlyProcessing, filePath)
	w.processingMu.Unlock()
}

func (w *Watcher) handleFileChange(ctx context.Context, filePath string) {
	// Check if already being processed (prevent concurrent processing of same file)
	if !w.markProcessing(filePath) {
		w.logger.Debug("file already being processed, skipping", "path", filePath)
		return
	}
	defer w.unmarkProcessing(filePath)

	w.logger.Info("processing file change", "path", filePath)

	// Check if file still exists
	_, err := os.Stat(filePath)
	fileExists := err == nil

	// Handle CONFIG.md specially
	if filepath.Base(filePath) == "CONFIG.md" {
		if fileExists {
			w.handleConfigChange(ctx, filePath)
		}
		return
	}

	// If file was deleted, remove chunks and hash
	if !fileExists {
		if err := w.db.DeleteChunksByFile(ctx, filePath); err != nil {
			w.logger.Error("failed to delete chunks", "path", filePath, "error", err)
		}
		w.clearFileHash(filePath)
		w.logger.Info("file deleted, chunks removed", "path", filePath)
		return
	}

	// Check if content has actually changed
	if !w.hasContentChanged(filePath) {
		w.logger.Debug("file content unchanged, skipping", "path", filePath)
		return
	}

	// Delete existing chunks for this file
	if err := w.db.DeleteChunksByFile(ctx, filePath); err != nil {
		w.logger.Error("failed to delete chunks", "path", filePath, "error", err)
	}

	// Parse the file
	chunks, err := w.parser.ParseFile(filePath)
	if err != nil {
		w.logger.Error("failed to parse file", "path", filePath, "error", err)
		return
	}

	if len(chunks) == 0 {
		// Update hash even if no chunks (file may be empty or non-indexable)
		w.updateFileHash(filePath)
		return
	}

	// Generate embeddings if enabled
	if w.db.EnableEmbeddings() && w.embedding != nil {
		contents := make([]string, len(chunks))
		for i, c := range chunks {
			contents[i] = c.Content
		}

		embeds, err := w.embedding.GenerateEmbeddingsBatch(ctx, contents, true)
		if err != nil {
			w.logger.Error("failed to generate embeddings", "path", filePath, "error", err)
		} else {
			for i := range chunks {
				chunks[i].Embedding = embeds[i]
			}
		}
	}

	// Index chunks
	if err := w.db.IndexChunks(ctx, chunks); err != nil {
		w.logger.Error("failed to index chunks", "path", filePath, "error", err)
		return
	}

	// Update file hash after successful indexing
	w.updateFileHash(filePath)

	w.logger.Info("reindexed file", "path", filePath, "chunks", len(chunks))
}

func (w *Watcher) handleConfigChange(ctx context.Context, filePath string) {
	content, err := os.ReadFile(filePath)
	if err != nil {
		w.logger.Error("failed to read CONFIG.md", "error", err)
		return
	}

	cfg, err := config.ParseConfigPage(string(content), true, w.denoRunner)
	if err != nil {
		w.logger.Error("failed to parse CONFIG.md", "error", err)
		return
	}

	if err := config.WriteConfigJSON(cfg, w.dbPath); err != nil {
		w.logger.Error("failed to write config JSON", "error", err)
		return
	}

	w.logger.Info("updated space config", "keys", len(cfg))
}

// InitialIndex performs a full index of the space.
func (w *Watcher) InitialIndex(ctx context.Context, rebuild bool) (int, error) {
	w.logger.Info("starting initial index", "rebuild", rebuild)

	if rebuild {
		if err := w.db.ClearDatabase(ctx); err != nil {
			return 0, err
		}
	}

	// Parse the space
	chunks, err := w.parser.ParseSpace(w.spacePath)
	if err != nil {
		return 0, err
	}

	w.logger.Info("parsed space", "chunks", len(chunks))

	// Index folders
	folderPaths, err := w.parser.GetFolderPaths(w.spacePath)
	if err != nil {
		return 0, err
	}

	indexPages, err := w.parser.GetFolderIndexPages(w.spacePath)
	if err != nil {
		return 0, err
	}

	if err := w.db.IndexFolders(ctx, folderPaths, indexPages); err != nil {
		return 0, err
	}

	w.logger.Info("indexed folders", "count", len(folderPaths))

	// Generate embeddings
	if w.db.EnableEmbeddings() && w.embedding != nil {
		contents := make([]string, len(chunks))
		for i, c := range chunks {
			contents[i] = c.Content
		}

		embeds, err := w.embedding.GenerateEmbeddingsBatch(ctx, contents, true)
		if err != nil {
			w.logger.Error("failed to generate embeddings", "error", err)
		} else {
			for i := range chunks {
				chunks[i].Embedding = embeds[i]
			}
		}
	}

	// Index chunks
	if err := w.db.IndexChunks(ctx, convertChunks(chunks)); err != nil {
		return 0, err
	}

	// Populate file hash cache for all indexed files
	w.hashMu.Lock()
	seenFiles := make(map[string]bool)
	for _, chunk := range chunks {
		if !seenFiles[chunk.FilePath] {
			seenFiles[chunk.FilePath] = true
			if hash, err := w.computeFileHash(chunk.FilePath); err == nil {
				w.fileHashes[chunk.FilePath] = hash
			}
		}
	}
	w.hashMu.Unlock()

	w.logger.Info("cached file hashes", "count", len(seenFiles))

	// Handle CONFIG.md
	configPath := filepath.Join(w.spacePath, "CONFIG.md")
	if content, err := os.ReadFile(configPath); err == nil {
		cfg, err := config.ParseConfigPage(string(content), true, w.denoRunner)
		if err == nil {
			_ = config.WriteConfigJSON(cfg, w.dbPath)
			w.logger.Info("wrote space config", "keys", len(cfg))
		}
	}

	w.logger.Info("initial index complete", "chunks", len(chunks))
	return len(chunks), nil
}

func convertChunks(parsed []types.Chunk) []types.Chunk {
	// Already the right type
	return parsed
}
